### PR TITLE
Reorginized common/types/util packages. Added response cache to block…

### PR DESCRIPTION
**Pull request by zalmen**
… validation (for ATXs and view)

### CODE REVIEW

File: common/types/encode.go [hunk 1]
```diff
@@ -63,6 +63,14 @@ func AtxAsBytes(tx *ActivationTx) ([]byte, error) {
 	return w.Bytes(), nil
 }
 
+func AtxIdsAsBytes(ids []AtxId) ([]byte, error) {
+	var w bytes.Buffer
```
Comment (almogdepaz) [inline common/types/encode.go]: we have an InterfaceToBytes method couldn't we just use that ?

File: sync/validation_queue.go [hunk 2]
```diff
@@ -101,9 +101,25 @@ func (vq *validationQueue) traverse(s *Syncer, blk *types.BlockHeader) error {
...
+			return fmt.Errorf("failed to check ATXs cache - err %v", err)
+		}
+		if atxsSeen == true && atxsRes == false {
+			return fmt.Errorf("invalid atxs according to cache (key %v)", atxsHash12)
+		}
+		txs, txErr, atxs, atxErr := s.DataAvailability(block, atxsSeen)
+		if txErr != nil || atxErr != nil {
+			if atxErr != nil {
+				// update cache only on failures, positive caching is done after all ATXs are processed
+				s.updateAtxCache(atxsHash12, false)
```
Comment (almogdepaz) [inline sync/validation_queue.go]: im not sure why we want to fail fetching atxs that we failed to fetch in the past, i dont think the failure to fetch necessarily suggests that the atxs are faulty

Comment (almogdepaz) [reply sync/validation_queue.go]: if you agree with me on this than we can just check the cache inside DataAvailabilty and avoid the code duplication here and in BlockSyntacticValidation

File: common/types/encode.go (hunk 1 above)
Comment (zalmen) [reply common/types/encode.go]: Fixed.
Now we always use `InterfaceToBytes` unless it is a slice of entities

File: sync/validation_queue.go (hunk 2 above)
Comment (zalmen) [reply sync/validation_queue.go]: per our discussion, removed both caches

**Review by almogdepaz — APPROVED** [review]

**Pull request merged**